README.md
# Exceptionally

[https://www.github.com/neilgupta/exceptionally](https://www.github.com/neilgupta/exceptionally)

[![Gem Version](https://badge.fury.io/rb/exceptionally.png)](http://badge.fury.io/rb/exceptionally)

Exceptionally abstracts your exception logic to make raising and returning meaningful errors in Ruby on Rails easier. It is primarily designed for API-only applications that need to return descriptive JSON error messages, rather than static HTML error pages.

Just raise the appropriate exception anywhere in your code and Exceptionally will handle the rest.

```ruby
raise Exceptionally::Unauthorized.new("Incorrect token")
```

will return the following JSON response to the client with a 401 status code: 

```javascript
{"error": "Incorrect token"}
```

It's that easy!

## Installation

To get started with Exceptionally, just add the following to your `Gemfile`:

```ruby
gem 'exceptionally'
```

Exceptionally has only been tested with Rails 4, but it should work with Rails 3+ in theory ;)

## Features

In addition to seamlessly abstracting your exception logic and returning clean JSON responses to your users, Exceptionally also offers the following benefits.

### Logging

If you raise a 500-level error, Exceptionally will log the error, backtrace, and relevant parameters to Rails.logger.

### Customizable

#### Add a custom error handler

Need to add more logging, integrate with Sentry, use [BacktraceCleaner](http://api.rubyonrails.org/classes/ActiveSupport/BacktraceCleaner.html), or do something else with the errors before they're returned to the user? Just add the following to `config/initializers/exceptionally.rb`:

```ruby
Exceptionally::Handler.before_render do |message, status, error, params|
  # Place your custom code here
  # message is a string description of what went wrong
  # status is an integer of the HTTP status code
  # error is the Exception object
  # params is a hash of the parameters passed to your controller

  # For example, you could:
  bc = BacktraceCleaner.new
  bc.add_filter   { |line| line.gsub(Rails.root, '') }
  bc.add_silencer { |line| line =~ /mongrel|rubygems/ }
  bc.clean(error.backtrace) # will strip the Rails.root prefix and skip any lines from mongrel or rubygems from your backtrace
end
```

#### Override output formatting

You can also override the returned response by adding a `render_error` method to your controller. For example, if you want to include the HTTP status code in the returned JSON, you can do:

```ruby
def render_error(error, status)
  render json: {error_message: error.message, error_status: status}, status: status
end
```

You could also return HTML, XML, or whatever other format is relevant to your application.

`render_error` must accept `error`, the StandardError object that was raised, and `status`, an integer HTTP status code.

#### Add custom errors

Exceptionally will handle the following errors by default:

* generic Exceptions
* ArgumentErrors
* ActiveRecord::RecordNotFound
* ActiveRecord::RecordInvalid
* ActiveRecord::RecordNotSaved
* ActionController::ParameterMissing
* Exceptionally errors (see below for available errors)

If there are additional errors that you want to assign status codes to and pass to Exceptionally, you can add the following to the top of your `application_controller.rb`:

```ruby
# Catch SomeGem::NotAuthorizedError errors and pass them to Exceptionally
rescue_from SomeGem::NotAuthorizedError, :with => :not_authorized_error

# Tell Exceptionally you want this treated as a 401 error
def not_authorized_error(error)
  pass_to_error_handler(error, 401, {caught_you: true})
end
```

`pass_to_error_handler` takes a Ruby Exception object, a status code, and an optional hash that will be returned in the json by merging it with the error message. If no status code is provided, it will default to 500.

## Available Errors

You can raise the following HTTP errors. All of them accept an optional string that is passed to the user instead of the generic message (eg. "Invalid token" instead of "Unauthorized"):

* Exceptionally::BadRequest
* Exceptionally::Unauthorized
* Exceptionally::PaymentRequired
* Exceptionally::Forbidden
* Exceptionally::NotFound
* Exceptionally::NotAllowed
* Exceptionally::NotAcceptable
* Exceptionally::ProxyAuthRequired
* Exceptionally::Timeout
* Exceptionally::Conflict
* Exceptionally::Gone
* Exceptionally::LengthRequired
* Exceptionally::PreconditionFailed
* Exceptionally::TooLarge
* Exceptionally::TooLong
* Exceptionally::UnsupportedMedia
* Exceptionally::RangeNotSatisfiable
* Exceptionally::ExpectationFailed
* Exceptionally::UnprocessableEntity
* Exceptionally::Error
* Exceptionally::NotImplemented
* Exceptionally::BadGateway
* Exceptionally::ServiceUnavailable
* Exceptionally::GatewayTimeout
* Exceptionally::HttpVersionNotSupported

You can also raise an error with just the HTTP status code by using `Exceptionally::Http404`. Status codes 400-417, 422, and 500-505 are available.

**[See descriptions of all status codes](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)**

## Why use Exceptionally?

By abstracting all of the exception handling logic, Exceptionally DRY's up your code and makes it easier to read. If you later decide to change the format of your error responses, you just need to edit `render_error` in one place. Exceptionally also transparently handles ActiveRecord and other generic exceptions for you, so that your app is less likely to crash. Additionally, you get a bunch of logging and error reporting functionality for free.

## Changelog

See [changelog](https://github.com/neilgupta/exceptionally/blob/master/CHANGELOG.md) to check for breaking changes between versions.

## Development

Run tests with `bundle exec rspec`

## Author

Neil Gupta [https://neil.gg](https://neil.gg)

## License

The MIT License (MIT) Copyright (c) 2023 Neil Gupta. See [MIT-LICENSE](https://raw.github.com/neilgupta/exceptionally/master/MIT-LICENSE)


CHANGELOG.md
# Exceptionally

[https://www.github.com/neilgupta/exceptionally](https://www.github.com/neilgupta/exceptionally)

## v1.5.0 (08/31/2023)

* Add Rails 7.0 support
* Add built-in rescue for ActiveRecord::RecordNotSaved
* Add built-in rescue for ActionController::ParameterMissing
* **Breaking**: Drop built-in error reporting to Raven, Airbrake, and New Relic
* **Breaking**: Drop built-in error handling of ApiPie errors

## v1.4.4 (03/22/2021)

* Add Rails 6.1 support

## v1.4.3 (08/28/2020)

* Add Rails 6 support

## v1.4.2 (02/11/2017)

* Add Rails 5 support

## v1.4.1 (06/29/2016)

* Filter sensitive request parameters using Rails' filter_parameters list

## v1.4.0 (12/15/2015)

* Remove built-in support for `Pundit::NotAuthorizedError` errors

## v1.3.0 (10/24/2015)

* Pass the whole error object to `render_error` rather than just the error message, allowing for more customization of what's rendered. This is a **breaking change** if you depend on the old `render_error` or `pass_to_error_handler` methods!

## v1.2.0 (10/14/2015)

* Add built-in support for `Pundit::NotAuthorizedError` errors
* Drop vestigial development dependencies

## v1.1.0 (06/27/2015)

* Add built-in support for reporting to Sentry
* Add config option to enable/disable built-in error reporting
* Add specs

## v1.0.1 (06/10/2015)

* Change ApiPie::ParamInvalid to return status code 422 rather than the generic 400 code

## v1.0.0 (03/16/2014)

* Initial release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

ENV['RAILS_ENV'] ||= 'test'

require 'rails'
require 'active_model'
require 'active_record/errors'
require 'action_view'
require 'action_controller'
require 'dummy/config/environment'
require 'rspec/rails'

require 'exceptionally'

RSpec.configure do |c|
  c.infer_base_class_for_anonymous_controllers = false
  c.order = 'random'
end


spec/dummy/version.rb
module Dummy
  VERSION = '0.0.1'
end


test/exceptionally_test.rb
require 'test_helper'

class ExceptionallyTest < ActiveSupport::TestCase
  test "truth" do
    assert_kind_of Module, Exceptionally
  end
end


lib/exceptionally/handler.rb
module Exceptionally
  class Handler
    def initialize(message = nil, status = nil, e = nil, params = nil)
      @message = message
      @status = status || 500
      @error = e
      @params = params || {}

      # ParameterFilter moved to ActiveSupport in rails 6
      filter = defined?(ActiveSupport::ParameterFilter) ? ActiveSupport::ParameterFilter : ActionDispatch::Http::ParameterFilter
      f = filter.new(Rails.application.config.filter_parameters)
      @params = f.filter @params

      @@callback.call(@message, @status, @error, @params) if defined?(@@callback) && @@callback.respond_to?(:call)

      log
    end

    def self.before_render(&block)
      @@callback = Proc.new(&block)
    end

    def log
      # Log 5xx errors
      if @status >= 500 && @error
        Rails.logger.error(@error.to_s)
        Rails.logger.error("Parameters: #{@params.to_s}") unless @params.blank?
        Rails.logger.error(@error.backtrace.join("\n")) unless @error.backtrace.blank?
      end
    end
  end
end


lib/exceptionally/controller.rb
require "active_record/validations.rb"

module Exceptionally
  module Controller
    extend ActiveSupport::Concern

    included do
      rescue_from Exception, :with => :exceptionally_handler
      rescue_from ArgumentError, :with => :exceptionally_handler
      rescue_from Exceptionally::Error, :with => :exceptionally_handler
      rescue_from ActiveRecord::RecordNotFound, :with => :missing_record_handler
      rescue_from ActiveRecord::RecordInvalid, :with => :record_invalid_handler
      rescue_from ActiveRecord::RecordNotSaved, :with => :record_not_saved
      rescue_from ActionController::ParameterMissing, :with => :missing_param_handler
    end

    # Raise custom error
    def exceptionally_handler(error)
      pass_to_error_handler(error)
    end

    # Raise 400 error
    def missing_param_handler(error)
      pass_to_error_handler(error, 400)
    end

    # Raise 404 error
    def missing_record_handler(error)
      pass_to_error_handler(error, 404)
    end

    # Raise 409 error
    def record_invalid_handler(error)
      pass_to_error_handler(error, 409)
    end

    # Raise 422 error
    def record_not_saved(error)
      pass_to_error_handler(error, 422, {validations: error.record.try(:errors)})
    end

    def pass_to_error_handler(error, status = nil, extra = {})
      status ||= error.try(:status) || 500
      Exceptionally::Handler.new(error.message, status, error, params)
      render_error(error, status, extra)
    end

    def render_error(error, status = 500, extra = {})
      render json: {error: error.message}.merge(extra || {}), status: status || 500
    end

  end
end


lib/exceptionally/exceptions.rb
module Exceptionally
  class Error < StandardError
    attr_reader :status

    def initialize(message = nil, status = nil, default = nil)
      @message = message
      @status = status || 500
      @default = default || "Internal Server Error"
    end

    def message
      @message || @default
    end

    def to_s
      "#{@status} #{@default}#{@message ? ": #{@message}" : ''}"
    end
  end

  class BadRequest < Error
    def initialize(message = nil)
      super(message, 400, "Bad Request")
    end
  end

  class Unauthorized < Error
    def initialize(message = nil)
      super(message, 401, "Unauthorized")
    end
  end

  class PaymentRequired < Error
    def initialize(message = nil)
      super(message, 402, "Payment Required")
    end
  end

  class Forbidden < Error
    def initialize(message = nil)
      super(message, 403, "Forbidden")
    end
  end

  class NotFound < Error
    def initialize(message = nil)
      super(message, 404, "Not Found")
    end
  end

  class NotAllowed < Error
    def initialize(message = nil)
      super(message, 405, "Method Not Allowed")
    end
  end

  class NotAcceptable < Error
    def initialize(message = nil)
      super(message, 406, "Not Acceptable")
    end
  end

  class ProxyAuthRequired < Error
    def initialize(message = nil)
      super(message, 407, "Proxy Authentication Required")
    end
  end

  class Timeout < Error
    def initialize(message = nil)
      super(message, 408, "Request Timeout")
    end
  end

  class Conflict < Error
    def initialize(message = nil)
      super(message, 409, "Conflict")
    end
  end

  class Gone < Error
    def initialize(message = nil)
      super(message, 410, "Gone")
    end
  end

  class LengthRequired < Error
    def initialize(message = nil)
      super(message, 411, "Length Required")
    end
  end

  class PreconditionFailed < Error
    def initialize(message = nil)
      super(message, 412, "Precondition Failed")
    end
  end

  class TooLarge < Error
    def initialize(message = nil)
      super(message, 413, "Request Entity Too Large")
    end
  end

  class TooLong < Error
    def initialize(message = nil)
      super(message, 414, "Request-URI Too Long")
    end
  end

  class UnsupportedMedia < Error
    def initialize(message = nil)
      super(message, 415, "Unsupported Media Type")
    end
  end

  class RangeNotSatisfiable < Error
    def initialize(message = nil)
      super(message, 416, "Requested Range Not Satisfiable")
    end
  end

  class ExpectationFailed < Error
    def initialize(message = nil)
      super(message, 417, "Expectation Failed")
    end
  end

  class UnprocessableEntity < Error
    def initialize(message = nil)
      super(message, 422, "Unprocessable Entity")
    end
  end

  class NotImplemented < Error
    def initialize(message = nil)
      super(message, 501, "Not Implemented")
    end
  end

  class BadGateway < Error
    def initialize(message = nil)
      super(message, 502, "Bad Gateway")
    end
  end

  class ServiceUnavailable < Error
    def initialize(message = nil)
      super(message, 503, "Service Unavailable")
    end
  end

  class GatewayTimeout < Error
    def initialize(message = nil)
      super(message, 504, "Gateway Timeout")
    end
  end

  class HttpVersionNotSupported < Error
    def initialize(message = nil)
      super(message, 505, "HTTP Version Not Supported")
    end
  end

  class Http400 < BadRequest; end
  class Http401 < Unauthorized; end
  class Http402 < PaymentRequired; end
  class Http403 < Forbidden; end
  class Http404 < NotFound; end
  class Http405 < NotAllowed; end
  class Http406 < NotAcceptable; end
  class Http407 < ProxyAuthRequired; end
  class Http408 < Timeout; end
  class Http409 < Conflict; end
  class Http410 < Gone; end
  class Http411 < LengthRequired; end
  class Http412 < PreconditionFailed; end
  class Http413 < TooLarge; end
  class Http414 < TooLong; end
  class Http415 < UnsupportedMedia; end
  class Http416 < RangeNotSatisfiable; end
  class Http417 < ExpectationFailed; end
  class Http422 < UnprocessableEntity; end
  class Http500 < Error; end
  class Http501 < NotImplemented; end
  class Http502 < BadGateway; end
  class Http503 < ServiceUnavailable; end
  class Http504 < GatewayTimeout; end
  class Http505 < HttpVersionNotSupported; end
end
